### PR TITLE
ast: Remove some indirection layers from values in key-value attributes

### PR DESCRIPTION
**Pull request by petrochenkov**
Trying to address some perf regressions from https://github.com/rust-lang/rust/pull/78837#issuecomment-745380762.

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -469,7 +469,9 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                         let count = generics
                             .params
                             .iter()
-                            .filter(|param| matches!(param.kind, ast::GenericParamKind::Lifetime { .. }))
```
Comment (Aaron1011) [inline compiler/rustc_ast_lowering/src/lib.rs:472]: Could you undo this change?

**Comment by Aaron1011** [comment]
@petrochenkov: Can you rebase against master?

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_ast_lowering/src/lib.rs:472]: Ok.
Apparently this is how nightly `rustfmt` works (from my editor), but `x.py fmt` doesn't change it back for some reason.

Comment (Mark-Simulacrum) [reply compiler/rustc_ast_lowering/src/lib.rs:472]: Probably we need to update the fmt used by x.py fmt, in stage0.txt. Filed https://github.com/rust-lang/rust/pull/80843 to do that.

**Comment by petrochenkov** [comment]
Rebased.

File: compiler/rustc_ast/src/ast.rs [hunk 2]
```diff
@@ -1478,10 +1478,10 @@ impl MacArgs {
     }
 
     pub fn span(&self) -> Option<Span> {
-        match *self {
+        match self {
             MacArgs::Empty => None,
             MacArgs::Delimited(dspan, ..) => Some(dspan.entire()),
-            MacArgs::Eq(eq_span, ref tokens) => Some(eq_span.to(tokens.span().unwrap_or(eq_span))),
+            MacArgs::Eq(_, token) => Some(token.span),
```
Comment (Aaron1011) [inline compiler/rustc_ast/src/ast.rs:1484]: Is it intentional that we're no longer including `eq_span`?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 3]
```diff
@@ -1013,10 +1013,36 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                        Some(TokenTree::Delimited(_, delim, tokens)) => {
+                            if delim != token::NoDelim {
+                                sess.diagnostic().delay_span_bug(
+                                    span,
+                                    "unexpected delimiter in key-value attribute's value",
+                                )
+                            }
+                            unwrap_single_token(sess, tokens, span)
+                        }
+                        None => Token::dummy(),
```
Comment (Aaron1011) [inline compiler/rustc_ast_lowering/src/lib.rs:1034]: Should there be a `delay_span_bug` here as well?

**Comment by Aaron1011** [comment]
r=me with the two comments addressed.

File: compiler/rustc_ast/src/ast.rs (hunk 2 above)
Comment (petrochenkov) [reply compiler/rustc_ast/src/ast.rs:1484]: No, this is a mistake, will fix.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_ast_lowering/src/lib.rs:1034]: This falls under `tokens.len() != 1`, so we already call `delay_span_bug` a few lines above in this case.

**Comment by rylev** [comment]
@petrochenkov @Aaron1011 It seems that this change caused some [perf regressions](https://perf.rust-lang.org/compare.html?start=6184f23950fb4aa14884ce310d948dc6fca269a3&end=f30733adb9150a9b965a6af1650d5213bec8ddf6&stat=instructions:u). Given that the perf run you ran above yielded only positive results, this is likely due to some other change getting in before this landed that introduced perf gains which this PR subsequently reversed. 

It looks like the largest perf hit was to the `expand_crate` query. 

Additionally, it's not clear to me how this change helped addressed the original perf regression as the impacted benchmarks were not improved by this PR. Would you mind clarifying that?

**Comment by petrochenkov** [comment]
I'm ok with reverting this PR if it causes mixed results, I wouldn't say it brings any improvements besides the perf effect.

>Additionally, it's not clear to me how this change helped addressed the original perf regression as the impacted benchmarks were not improved by this PR. Would you mind clarifying that?

https://github.com/rust-lang/rust/pull/78837 introduced more indirection by boxing literals in interpolated expressions, this PR removed some indirection by keeping a single token instead of a `Vec` of tokens.
So these two indirection are located in different places, and may affect different tests differently, but I still hoped for them to counter-balance each other on average.

The ultimate solution to the regressions from https://github.com/rust-lang/rust/pull/78837 is to keep AST on arena (like it's already done with HIR), so that *no* expressions are ever boxed. That would help not only expressions in key-value attributes, but all expressions in functions as well (which are in majority).
